llama-index-packs/llama-index-packs-query-understanding-agent/llama_index/packs/query_understanding_agent/step.py

```diff
@@ -0,0 +1,281 @@
+) -> ChatPromptTemplate:
+    system_msg = ChatMessage(role=MessageRole.SYSTEM, content=system_prompt)
+    messages = [system_msg]
+    for raw_msg in current_reasoning:
+        if raw_msg[0] == "user":
+            messages.append(ChatMessage(role=MessageRole.USER, content=raw_msg[1]))
+        else:
+            messages.append(ChatMessage(role=MessageRole.ASSISTANT, content=raw_msg[1]))
+    return ChatPromptTemplate(message_templates=messages)
+
+
+class ResponseEval(BaseModel):
+    """Evaluation of whether the response has an error."""
+
+    clarifying_question: str = Field(
+        ..., description="The clarifying question, if human input is required"
+    )
+    explanation: str = Field(
+        ...,
+        description=(
+            "The explanation for the error OR for the clarifying question."
+            "Can include the direct stack trace as well."
+        ),
+    )
+    has_error: bool = Field(..., description="Whether the response has an error")
+    requires_human_input: bool = Field(
+        ...,
+        description="Whether the response needs human input. If true, the clarifying question should be provided.",
+    )
+
+
+class HumanInputRequiredException(Exception):
+    """Exception raised when human input is required."""
+
+    def __init__(
+        self,
+        message="Human input is required",
+        task_id: Optional[str] = None,
+        step: TaskStep = None,
+    ):
+        self.message = message
+        self.task_id = task_id
+        self.step = step
+        super().__init__(self.message)
+
+
+class QueryUnderstandingAgentWorker(CustomSimpleAgentWorker):
+    """Agent worker that adds a retry layer on top of a router.
+
+    Continues iterating until there's no errors / task is done.
+
+    """
+
+    prompt_str: str = Field(default=DEFAULT_PROMPT_STR)
+    max_iterations: int = Field(default=10)
+
+    _router_query_engine: RouterQueryEngine = PrivateAttr()
+
+    def __init__(self, tools: List[BaseTool], **kwargs: Any) -> None:
+        """Init params."""
+        # validate that all tools are query engine tools
+        for tool in tools:
+            if not isinstance(tool, QueryEngineTool):
+                raise ValueError(
+                    f"Tool {tool.metadata.name} is not a query engine tool."
+                )
+        self._router_query_engine = RouterQueryEngine(
+            selector=PydanticSingleSelector.from_defaults(),
+            query_engine_tools=tools,
+            verbose=kwargs.get("verbose", False),
+        )
+        super().__init__(
+            tools=tools,
+            **kwargs,
+        )
+
+    def _initialize_state(self, task: Task, **kwargs: Any) -> Dict[str, Any]:
+        """Initialize state."""
+        return {"count": 0, "current_reasoning": []}
+
+    def _run_llm_program(self, prompt, state):
+        for _ in range(3):
+            try:
+                return LLMTextCompletionProgram.from_defaults(
```

Comment:
We can actually use `llm.structured_predict` to make the selection between pydantic program and llm program automatic

Reply:
something like this


```
        response_eval = self.llm.structured_predict(
            ResponseEval,
            PromptTemplate(self.prompt_str),
            query_str=new_input,
            response_str=str(response),
            tools=tools,
        )
```